Userland/Services/SystemServer/main.cpp

```diff
@@ -520,6 +210,82 @@ static ErrorOr<void> create_tmp_semaphore_directory()
     return {};
 }
 
+static ErrorOr<void> reopen_base_file_descriptors()
+{
+    // Note: We open the /dev/null device and set file descriptors 0, 1, 2 to it
+    // because otherwise these file descriptors won't have a custody, making
+    // the ProcFS file descriptor links (at /proc/PID/fd/{0,1,2}) to have an
+    // absolute path of "device:1,3" instead of something like "/dev/null".
+    // This affects also every other process that inherits the file descriptors
+    // from SystemServer, so it is important for other things (also for ProcFS
+    // tests that are running in CI mode).
+    int stdin_new_fd = TRY(Core::System::open("/dev/null"sv, O_NONBLOCK));
+    TRY(Core::System::dup2(stdin_new_fd, 0));
+    TRY(Core::System::dup2(stdin_new_fd, 1));
+    TRY(Core::System::dup2(stdin_new_fd, 2));
+    return {};
+}
+
+static ErrorOr<void> activate_services(Core::ConfigFile const& config)
+{
+    for (auto const& name : config.groups()) {
+        auto service = TRY(Service::try_create(config, name));
+        TRY(service->setup_sockets());
+        if (service->is_enabled_for_system_mode(g_system_mode))
+            g_services.append(move(service));
+    }
+    // After we've set them all up, activate them!
+    dbgln("Activating {} services...", g_services.size());
+    for (auto& service : g_services) {
+        dbgln_if(SYSTEMSERVER_DEBUG, "Activating {}", service->name());
+    }
+
+    for (auto& service : g_services) {
+        if (auto result = service->activate(); result.is_error())
+            dbgln("{}: {}", service->name(), result.release_error());
+    }
+    return {};
+}
+
+static ErrorOr<void> activate_base_services_based_on_system_mode()
+{
+    pid_t pid = TRY(Core::System::fork());
+    if (pid == 0) {
+        bool found_gpu_device = false;
+        for (int attempt = 0; attempt < 10; attempt++) {
+            struct stat file_state;
+            int rc = lstat("/dev/gpu/connector0", &file_state);
+            if (rc == 0) {
+                found_gpu_device = true;
+                break;
+            }
+            sleep(1);
+        }
+        if (!found_gpu_device) {
+            dbgln("WARNING: No device nodes at /dev/gpu/ directory after 10 seconds. This is probably a sign of disabled graphics functionality.");
+            dbgln("To cope with this, graphical mode will not be enabled.");
+            g_system_mode = "text";
+        }
```

Comment:
Looks like we now hit this timeout on CI. Missing a check for `g_system_mode == "graphical"sv` before entering the loop?